File: src/offline_trigger.cpp
#include <iostream>
#include <sstream>
#include <string>
#include <boost/program_options.hpp>
#include "OfflineTrigger.h"

namespace po = boost::program_options;

int main(int argc, const char **argv) {

    po::options_description desc("Options");
    desc.add_options()
        ("help", "produce help message")
        ("prefix", po::value<std::string>(), "prefix for saved files")
        ("run_n", po::value<int>(), "run number")
        ("save_every", po::value<int>(), "save every n frames, even if a photon is not detected")
        ("min_pedestal_queue", po::value<int>(), "minimum number of frames to be averaged for the calculation of the pedestal")
        ("really_delete_files", po::value<bool>(), "really delete original files")
        ;
    po::variables_map vm;
    po::store(po::parse_command_line(argc, argv, desc), vm);
    po::notify(vm);

    std::string example = "./offline_trigger --prefix=/afs/psi.ch/project/hedpc/raw_data/gotthard/2013.01.24/Nd_ --run_n=0 --save_every=1000 --min_pedestal_queue=1000 --really_delete_files=0";

    if (vm.count("help")) {
        std::cout << desc << std::endl;
        std::cout << example << std::endl;
        return 1;
    }

    std::string prefix;
    int run_n, save_every, min_pedestal_queue;
    bool really_delete_files;
    if (vm.count("prefix") and
            vm.count("run_n") and
            vm.count("save_every") and
            vm.count("min_pedestal_queue") and
            vm.count("really_delete_files")) {
        prefix = vm["prefix"].as<std::string>();
        run_n = vm["run_n"].as<int>();
        save_every = vm["save_every"].as<int>();
        min_pedestal_queue = vm["min_pedestal_queue"].as<int>();
        really_delete_files = vm["really_delete_files"].as<bool>();
        std::cout << prefix << " " << run_n << " " << save_every << " " << min_pedestal_queue << " " << really_delete_files << std::endl;
    }
    else {
        std::cerr << "options missing!" << std::endl;
        std::cerr << desc << std::endl;
        std::cerr << example << std::endl;
        return 2;
    }

    OfflineTrigger trigger(prefix, run_n, save_every, min_pedestal_queue, really_delete_files);
    int saved_frames = trigger.start_triggering();
    std::cout << "saved a total of " << saved_frames << " frames." << std::endl;
    return 0;
}

File: python/offline_trigger.py
#!/usr/bin/env python
from __future__ import division, print_function

import argparse
import os
parser = argparse.ArgumentParser(description='''analyze spectrum recorded by gotthard.
        example:
        python offline_trigger.py /afs/psi.ch/project/hedpc/raw_data/gotthard/2013.01.18/Ir_ 0 1000 1000
''')
parser.add_argument('prefix', metavar='PREFIX',
        nargs=1, help='prefix of the files containing the data')
parser.add_argument('run_number', type=int, metavar='RUN_N',
        nargs=1, help='run number')
parser.add_argument('save_every', type=int, metavar='SAVE_EVERY',
        nargs=1, help='save every SAVE_EVERY frames, even if a photon is not detected')
parser.add_argument('min_pedestal_size', type=int, metavar='MIN_PEDESTAL',
        nargs=1, help='minimum size of the pedestal queue')
args = parser.parse_args()

prefix = args.prefix[0]
run_number = args.run_number[0]
save_every = args.save_every[0]
min_pedestal_size = args.min_pedestal_size[0]

import ROOT
from rootstyle import tdrstyle

tdrstyle()
ROOT.gROOT.ProcessLine(".L gotthard_utils.cpp+")
ROOT.gROOT.ProcessLine(".L Packet.cpp+")
ROOT.gROOT.ProcessLine(".L FullFrame.cpp+")
ROOT.gROOT.ProcessLine(".L FrameROOTFunctions.cpp+")
ROOT.gROOT.ProcessLine(".L PedestalCalculator.cpp+")
ROOT.gROOT.ProcessLine(".L OfflineTrigger.cpp+")

trigger = OfflineTrigger(prefix, run_number, save_every, min_pedestal_size)
trigger.start_triggering()
